common/src/main/java/tachyon/resource/ResourcePool.java

```diff
@@ -66,19 +67,33 @@ protected ResourcePool(int maxCapacity, BlockingQueue<T> resources) {
    */
   public T acquire() {
     // If the resource pool is empty but capacity is not yet full, create a new resource.
-    synchronized (mCapacityLock) {
-      if (mResources.isEmpty() && mCurrentCapacity < mMaxCapacity) {
-        T newResource = createNewResource();
-        mCurrentCapacity ++;
-        return newResource;
-      }
+    T resource = mResources.poll();
+    if (resource != null) {
+      return resource;
+    }
+
+    if (mCurrentCapacity.getAndIncrement() < mMaxCapacity) {
+      return createNewResource();
     }
 
+    mCurrentCapacity.decrementAndGet();
+
     // Otherwise, try to take a resource from the pool, blocking if none are available.
     try {
-      return mResources.take();
+      mTakeLock.lockInterruptibly();
+      try {
+        while (true) {
+          resource = mResources.poll();
+          if (resource != null) {
+            mNotEmpty.signal();
```

Comment:
I don't think I understand. The signal() is used to notify other threads waiting with mNotEmpty.await()?

If so, why would taking a resource need to signal other waiting threads? Wouldn't only release() need to signal() those threads?